File: 剑指Offer/cpp/src/_51_数组中的逆序对.cpp
#include <vector>
using namespace std;
// todo 数状数组
class UsingMergeSort
{
public:
    int reversePairs(vector<int> &nums)
    {
        vector<int> cache(nums.size());
        return merge(nums, 0, nums.size() - 1, cache);
    }

    int merge(vector<int> &nums, int left, int right, vector<int> &cache)
    {
        if (left >= right)
            return 0;
        int mid = left + ((right - left) >> 1);
        int ret = merge(nums, left, mid, cache);
        ret += merge(nums, mid + 1, right, cache);

        int i = mid, j = right, k = j;
        while (i >= left)
        {
            while (j > mid && nums[i] <= nums[j])
            {
                cache[k--] = nums[j--];
            }
            ret += j - mid;
            cache[k--] = nums[i--];
        }
        while (j > mid)
        {
            cache[k--] = nums[j--];
        }
        for (k = left; k <= right; k++)
        {
            nums[k] = cache[k];
        }
        return ret;
    }
};

File: 剑指Offer/kotlin/src/_51_数组中的逆序对.kt
class _51_数组中的逆序对_ {
    class Solution {
        fun reversePairs(nums: IntArray): Int {
            val n = nums.size
            val cache = IntArray(n)
            fun merge(left: Int, right: Int): Int {
                if (left >= right) return 0
                val mid = (left + right) ushr 1
                var ret = merge(left, mid)
                ret += merge(mid + 1, right)

                var i = mid
                var j = right
                var k = right
                while (i >= left) {
                    while (j > mid && nums[i] <= nums[j]) {
                        cache[k--] = nums[j--]
                    }
                    ret += j - mid
                    cache[k--] = nums[i--]
                }
                while (j > mid) {
                    cache[k--] = nums[j--]
                }
                cache.copyInto(nums, left, left, right + 1)
                return ret
            }
            return merge(0, n -1)
        }
    }
}